Description: Fix prototype pollution in minimist argument parsing (CVE-2021-44906).
 The setKey() walker followed arbitrary key paths (e.g. --__proto__.x,
 --x.__proto__.z, --constructor.prototype.y) allowing an attacker to mutate
 Object.prototype and other built-in prototypes. Guard every intermediate key
 and the final key against __proto__ / constructor-prototype traversal, and
 refuse to write onto built-in prototype objects. Adapted from upstream
 minimist commits ef9153f / 34e20b8.
Origin: backport, https://github.com/minimistjs/minimist/commit/ef9153fc52b6cea0744b2239921c5dcae4697f11
--- a/index.js
+++ b/index.js
@@ -157,21 +157,34 @@ function hasKey (obj, keys) {
     return key in o;
 }

+function isConstructorOrProto (obj, key) {
+    return key === 'constructor' && (typeof obj[key] === 'function' || key === '__proto__');
+}
+
 function setKey (obj, keys, value) {
     var o = obj;
-    keys.slice(0,-1).forEach(function (key) {
-        if (o[key] === undefined) o[key] = {};
-        o = o[key];
-    });
-    
+    for (var i = 0; i < keys.length - 1; i++) {
+        var ikey = keys[i];
+        if (ikey === '__proto__' || isConstructorOrProto(o, ikey)) return;
+        if (o[ikey] === undefined) o[ikey] = {};
+        if (o[ikey] === Object.prototype || o[ikey] === Number.prototype
+        || o[ikey] === String.prototype) o[ikey] = {};
+        if (o[ikey] === Array.prototype) o[ikey] = [];
+        o = o[ikey];
+    }
+
     var key = keys[keys.length - 1];
+    if (key === '__proto__' || isConstructorOrProto(o, key)) return;
+    if (o === Object.prototype || o === Number.prototype
+    || o === String.prototype) o = {};
+    if (o === Array.prototype) o = [];
     if (o[key] === undefined || typeof o[key] === 'boolean') {
         o[key] = value;
     }
     else if (Array.isArray(o[key])) {
         o[key].push(value);
     }
     else {
         o[key] = [ o[key], value ];
     }
 }
--- a/test/proto.js
+++ b/test/proto.js
@@ -0,0 +1,46 @@
+var parse = require('../');
+var test = require('tape');
+
+/* eslint no-proto: 0 */
+
+test('proto pollution', function (t) {
+    var argv = parse(['--__proto__.x', '123']);
+    t.equal({}.x, undefined);
+    t.equal(argv.__proto__.x, undefined);
+    t.equal(argv.x, undefined);
+    t.end();
+});
+
+test('proto pollution (array)', function (t) {
+    var argv = parse(['--x', '4', '--x', '5', '--x.__proto__.z', '789']);
+    t.equal({}.z, undefined);
+    t.deepEqual(argv.x, [4, 5]);
+    t.equal(argv.x.z, undefined);
+    t.equal(argv.x.__proto__.z, undefined);
+    t.end();
+});
+
+test('proto pollution (number)', function (t) {
+    var argv = parse(['--x', '5', '--x.__proto__.z', '100']);
+    t.equal({}.z, undefined);
+    t.equal((4).z, undefined);
+    t.equal(argv.x, 5);
+    t.equal(argv.x.z, undefined);
+    t.end();
+});
+
+test('proto pollution (string)', function (t) {
+    var argv = parse(['--x', 'abc', '--x.__proto__.z', 'def']);
+    t.equal({}.z, undefined);
+    t.equal('...'.z, undefined);
+    t.equal(argv.x, 'abc');
+    t.equal(argv.x.z, undefined);
+    t.end();
+});
+
+test('proto pollution (constructor)', function (t) {
+    var argv = parse(['--constructor.prototype.y', '123']);
+    t.equal({}.y, undefined);
+    t.equal(argv.y, undefined);
+    t.end();
+});
